server.py

```diff
@@ -18,8 +19,10 @@ class Server:
     def __init__(self, imagedata):
         self.board = CanvasBoard(imagedata)
         self.lock = threading.Lock()
-    def updateBoard(self,imagedata):
-        self.board = CanvasBoard(imagedata)
+    def updateBoard(self, diff_as_dict):
+        numPixels = len(diff_as_dict['coord'])
+        for i in range(numPixels):
+            self.board.data[diff_as_dict['coord'][i]] = diff_as_dict['val'][i]

```

Comment:
Let's abstract this away into a member function in CanvasBoard:

Also don't forget to **grab** the lock _before_ manipulating the shared resource and **release** the lock _afterward_.

```
def updateBoard(self, diff: dict):
  self.lock.acquire()
  canvasBoard.updateWithDiff(diff)
  self.lock.release()
```

Reply:
Updating board now abstracted into CanvasBoard class as CanvasBoard.updateBoard(). Lock and release added for mutex.